internal/datacoord/dataview/view_manager.go

```diff
@@ -0,0 +1,176 @@
+	}
+}
+
+func (m *dataViewManager) Start() {
+	ticker := time.NewTicker(paramtable.Get().DataCoordCfg.DataViewUpdateInterval.GetAsDuration(time.Second))
+	defer ticker.Stop()
+	for {
+		select {
+		case <-m.closeChan:
+			log.Info("data view manager exited")
+			return
+		case <-ticker.C:
+			// periodically update all data view
+			for _, collectionID := range m.currentViews.Keys() {
+				m.TryUpdateDataView(collectionID)
+			}
+		case collectionID := <-updateChan:
+			m.TryUpdateDataView(collectionID)
+		}
+	}
+}
+
+func (m *dataViewManager) Close() {
+	m.closeOnce.Do(func() {
+		close(m.closeChan)
+	})
+}
+
+func (m *dataViewManager) update(view *DataView, reason string) {
+	m.currentViews.Insert(view.CollectionID, view)
+	log.Info("update new data view", zap.Int64("collectionID", view.CollectionID), zap.Int64("version", view.Version), zap.String("reason", reason))
+}
+
+func (m *dataViewManager) TryUpdateDataView(collectionID int64) {
+	newView, err := m.pullFn(collectionID)
+	if err != nil {
+		log.Warn("pull new data view failed", zap.Int64("collectionID", collectionID), zap.Error(err))
+		// notify to trigger retry
+		NotifyUpdate(collectionID)
+		return
+	}
+
+	currentView, ok := m.currentViews.Get(collectionID)
+	if !ok {
+		// update due to data view is empty
+		m.update(newView, "init data view")
+		return
+	}
+	// no-op if the incoming version is less than the current version.
+	if newView.Version <= currentView.Version {
+		log.Warn("stale version, skip update", zap.Int64("collectionID", collectionID),
+			zap.Int64("new", newView.Version), zap.Int64("current", currentView.Version))
+		return
+	}
+
+	for channel, new := range newView.Channels {
+		current, ok := currentView.Channels[channel]
+		if !ok {
+			// update due to channel info is empty
+			m.update(newView, "init channel info")
+			return
+		}
+		if !funcutil.SliceSetEqual(new.GetLevelZeroSegmentIds(), current.GetLevelZeroSegmentIds()) ||
+			!funcutil.SliceSetEqual(new.GetUnflushedSegmentIds(), current.GetUnflushedSegmentIds()) ||
+			!funcutil.SliceSetEqual(new.GetFlushedSegmentIds(), current.GetFlushedSegmentIds()) ||
+			!funcutil.SliceSetEqual(new.GetIndexedSegmentIds(), current.GetIndexedSegmentIds()) ||
+			!funcutil.SliceSetEqual(new.GetDroppedSegmentIds(), current.GetDroppedSegmentIds()) {
+			// update due to segments list changed
+			m.update(newView, "channel segments list changed")
+			return
+		}
+		if !typeutil.MapEqual(new.GetPartitionStatsVersions(), current.GetPartitionStatsVersions()) {
+			// update due to partition stats changed
+			m.update(newView, "partition stats changed")
+			return
+		}
+		newTime := tsoutil.PhysicalTime(new.GetSeekPosition().GetTimestamp())
+		curTime := tsoutil.PhysicalTime(current.GetSeekPosition().GetTimestamp())
+		if newTime.Sub(curTime) > paramtable.Get().DataCoordCfg.CPIntervalToUpdateDataView.GetAsDuration(time.Second) {
```

Comment:
update the view every 10 minutes as a fallback polity

Reply:
done